File: DataStructuresJava/src/main/java/com/shaynek/datastruct/kt/LinkedList.java
package com.shaynek.datastruct.kt;

public class LinkedList<T> {

    private int length;
    private Node<T> head;

    public LinkedList() {
        length = 0;
        head = null;
    }

    @Override
    public String toString() {
        if (head == null) return "Empty linked list";

        Node<T> current = head;
        StringBuilder builder = new StringBuilder();

        while (current != null) {
            final T data = current.getData();
            builder.append(data != null ? data.toString() : "null")
                    .append(" -> ");
            current = current.next;
        }
        builder.append("null");
        return builder.toString();
    }

    public T first() {
        return head != null ? head.getData() : null;
    }

    public T last() {
        return lastNode() != null ? lastNode().getData() : null;
    }

    public void append(T data) {
        final Node<T> newNode = new Node<>(data);
        length++;
        if (head == null) {
            head = newNode;
        } else if (lastNode() != null) {
            lastNode().next = newNode;
        }
    }

    public int getLength() {
        return length;
    }

    private Node<T> lastNode() {
        if (head == null) return null;
        Node<T> current = head;
        while (current.next != null) {
            current = current.next;
        }
        return current;
    }

    private class Node<V> {
        private V data;
        private Node<V> next;

        Node(V data) {
            this.data = data;
            next = null;
        }

        V getData() {
            return data;
        }
    }
}

File: DataStructuresKotlin/src/main/kotlin/com/shaynek/datastruct/LinkedList.kt
package com.shaynek.datastruct

class LinkedList<T> {

    var length = 0
    private var head: Node<T>? = null

    override fun toString(): String {
        if (head == null) return "Empty linked list"

        var current = head
        val builder = StringBuilder()

        while (current != null) {
            val dataString = current.data?.toString() ?: "null"
            builder.append("$dataString -> ")
            current = current.next
        }
        builder.append("null")
        return builder.toString()
    }

    fun first(): T? = head?.data
    fun last(): T? = lastNode()?.data

    fun append(data: T) {
        length++
        Node(data).also {
            if (head == null) head = it
            else lastNode()?.next = it
        }
    }

    private fun lastNode(): Node<T>? {
        var current = head
        while (current?.next != null) {
            current = current.next
        }
        return current
    }
}

private data class Node<T>(val data: T) {
    var next: Node<T>? = null
}